parsl/monitoring/file_monitoring.py

```diff
@@ -0,0 +1,488 @@
+        or ``filetype`` must be given.
+    path: str, optional
+        The base path where the files are expected to be, default is current working directory (``None``).
+    working_dir: str, optional
+        The working directory for the file monitoring, default (``None``) is the current working directory
+        inherited when the monitor process is started.
+    email: str, optional
+        Email address to which outputs from the callbacks are sent. Default is None
+    sleep_dur: float
+        The time to wait between scans of the file system to look for matching files. Default is 60 seconds.
+    max_proc: int, optional
+        The maximum size of the multiprocessing Pool for the file processing. Default is 4, but the
+        value is adjusted to be ``min(max_proc, len(patterns))`` so as to not make the pool larger than
+        needed.
+    """
+    def __init__(self,
+                 callback: Union[Callable, List[Callable]],
+                 pattern: Optional[Union[str, List[str]]] = None,
+                 filetype: Optional[Union[str, List[str]]] = None,
+                 path: Optional[str] = None,
+                 working_dir: Optional[str] = None,
+                 email: Optional[str] = None,
+                 sleep_dur: float = 60.,
+                 max_proc: int = 4):
+        logger.info(f"File_monitor initialized: {path}, {working_dir}, {email}, {sleep_dur}")
+        self.email = email
+        # generate the master pattern list
+        self.patterns: List[Tuple[Any, bool]] = []
+        if pattern is not None:
+            if isinstance(pattern, list):
+                for p in pattern:
+                    self.patterns.append((re.compile(p), True))
+            else:
+                self.patterns = [(re.compile(pattern), True)]
+        if filetype is not None:
+            temppat = []
+            if isinstance(filetype, list):
+                temppat = filetype
+            else:
+                temppat.append(filetype)
+            if path is not None:
+                for i, pat in enumerate(temppat):
+                    if '*' not in pat:
+                        if not pat.startswith('.'):
+                            pat = '.' + pat
+                        pat = '*' + pat
+                    temppat[i] = os.path.join(path, pat)
+            else:
+                for i, pat in enumerate(temppat):
+                    if '*' not in pat:
+                        if not pat.startswith('.'):
+                            pat = '.' + pat
+                        pat = '*' + pat
+                    temppat[i] = pat
+            for ft in temppat:
+                self.patterns.append((ft, False))
+        if not self.patterns:
+            raise Exception("Either pattern or filetype must be given")
+        if isinstance(callback, list):
+            if len(self.patterns) != len(callback):
+                raise Exception("Pattern list is not the same size as function list")
```

Comment:
there's a different style you could maybe use that makes this true by construction:

instead of three lists of callback, pattern, filetype, you instead have one list, with the contents being a tuple: (callback, pattern, filetype)

that's maybe also easier for users when they edit their code - instead of three separate lists they need to keep aligned manually, there's a single list and inserting an element in that list does not break the other two.